x/programs/runtime/utils.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package runtime
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/near/borsh-go"
+)
+
+func EnsureInt64ToInt32(v int64) bool {
+	return v >= math.MinInt32 && v <= math.MaxInt32
+}
+
+func EnsureUint32ToInt32(v uint32) bool {
+	return v <= math.MaxInt32
+}
+
+// SmartPtr is an int64 where the first 4 bytes represent the length of the bytes
+// and the following 4 bytes represent a pointer to WASM memeory where the bytes are stored.
+type SmartPtr int64
+
+// ptrMask is used to get the lower 32 bits of a SmartPtr
+var ptrMask = ^uint32(0)
+
+// FromSmartPtr returns the bytes at [ptr] in [client] memory.
+func FromSmartPtr(client WasmtimeExportClient, smartPtr SmartPtr) ([]byte, error) {
+	memory := NewMemory(client)
+
+	ptr, length := splitSmartPtr(smartPtr)
+
+	// The following [length] bytes represent the bytes to return
+	bytes, err := memory.Range(uint64(ptr), uint64(length))
+	if err != nil {
+		return nil, err
+	}
+
+	return bytes, nil
+}
+
+// ToSmartPtr converts [ptr] and [len] to a SmartPtr.
+func ToSmartPtr(ptr uint32, len uint32) (SmartPtr, error) {
+	// ensure length of bytes is not greater than int32 to prevent overflow
+	if !EnsureUint32ToInt32(len) {
+		return 0, fmt.Errorf("length of bytes is greater than int32")
+	}
+
+	// convert to int64 with length of bytes in the upper 32 bits and ptr in the lower 32 bits
+	smartPtr := int64(len) << 32
+	smartPtr |= int64(ptr)
```

Comment:
optional. I think we could break this up a little to make it easier to understand the bitwise op.
```go
lenUpperBits := int64(len) << 32
ptrLowerBits := int64(ptr)

smartPtr := lenUpperBits | ptrLowerBits
```